text/0000-futures.md

```diff
@@ -0,0 +1,811 @@
+number of basic adapters (described in the reference below), while some
+particularly interesting variants are iterating in the crates.io ecosystem
+first.
+
+Ultimately asynchronous computations are executed by *tasks*, which are
+lightweight threads. In particular, an *executor* is able to "spawn" a
+`()`-producing future as an independent task; these tasks are then cooperatively
+scheduled onto one or more operating system threads. The `Executor` trait
+defines this interface, and the `task` module provides a host of related
+definitions needed when manually implementing futures or executors.
+
+*Note: additional guide-level documentation is available in
+the [futures crate](https://docs.rs/futures/0.2.0-beta/futures/)*.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## `core::task` module
+
+The fundamental mechanism of futures is *tasks*, which are lightweight
+threads of execution; many tasks can be cooperatively scheduled onto a single
+operating system thread.
+
+To perform this cooperative scheduling we use a technique sometimes referred to
+as a "trampoline". When a task would otherwise need to block waiting for some
+event, instead it schedules itself for later wakeup and *returns* to the
+executor running it, which can then run another task. Subsequent wakeups place
+the task back on the executors queue of ready tasks, much like a thread
+scheduler in an operating system.
+
+Attempting to complete a task (or future within it) is called *polling*, and
+always yields a `Poll` value back:
+
+```rust
+/// Indicates whether a value is available, or if the current task has been
+/// scheduled for later wake-up instead.
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub enum Poll<T> {
+    /// Represents that a value is immediately ready.
+    Ready(T),
+
+    /// Represents that a value is not ready yet.
+    ///
+    /// When a function returns `Pending`, the function *must* also
+    /// ensure that the current task is scheduled to be awoken when
+    /// progress can be made.
+    Pending,
+}
+```
+
+When a task returns `Poll::Ready`, the executor knows the task has completed and
+can be dropped.
+
+### Waking up
+
+Each task executor provides its own scheduling facilities, and hence needs to
+customize the way task wakeups are handled. As such, there is a
+`std::task::Wake` trait defining wakeup behavior:
+
+```rust
+/// A way of waking up a specific task.
+///
+/// Any task executor must provide a way of signaling that a task it owns
+/// is ready to be `poll`ed again. Executors do so by implementing this trait.
+pub trait Wake: Send + Sync {
+    /// Indicates that the associated task is ready to make progress and should
+    /// be `poll`ed.
+    ///
+    /// Executors generally maintain a queue of "ready" tasks; `wake` should place
+    /// the associated task onto this queue.
+    fn wake(&Arc<self>);
+}
+```
```

Comment:
After reading the RFC a couple more time, I think I have finally reverse-engineered in my mind why one would want to use an Arc here.

1. Context only takes a shared reference to a Waker, so all it can provide to the executing future is a shared reference.
2. But the future may potentially need to be awoken from a different, asynchronous code path, in which case there is a need for an owned Waker value that can be moved around.
3. There should thus be a way to clone the Waker and potentially send the output to a different thread, ergo, we need an Arc.

Of those, only point 1 is somewhat controversial. We could have imagined an alternate design in which the Context owns a Waker, and one can get it by consuming the Context (after all, do you still need that context if you are going to suspend the future).

But this means that a new Waker must be created every time a future is polled, even if it is unused, and atomic increments/decrements are not cheap so we don't want to do that. Alternatively, we might provide Context with a way to _create_ a Waker, but that would ultimately get quite complicated.

So considering that there are use cases for futures that can be awoken from multiple places (anything that combines multiple futures into one: join(), select()...), an Arc was probably deemed to be the most appropriate implementation.

Definitely not obvious, and I agree with you that this RFC would benefit from explaining a bit more the "why", rather than mostly the "what".